lib/dotcom/system_status/groups.ex

```diff
@@ -0,0 +1,221 @@
+      %{} -> true
+    end)
+  end
+
+  defp add_now_text(%{time: nil} = status), do: %{status | time: "Now"}
+  defp add_now_text(status), do: status
+
+  defp consolidate_duplicate_descriptions(statuses) do
+    statuses
+    |> Enum.group_by(fn %{time: time, description: description} -> {time, description} end)
+    |> Enum.map(fn
+      {_, [status]} -> status
+      {_, [status | _]} -> pluralize_description(status)
+    end)
+  end
+
+  defp pluralize_description(%{description: description} = status) do
+    new_description =
+      case description do
+        "Suspension" -> "Suspensions"
+        "Station Closure" -> "Station Closures"
+        _ -> description
+      end
+
+    %{status | description: new_description}
+  end
+
+  defp combine_green_line_branches(statuses_by_route) do
+    {green_line_entries, other_entries} =
+      statuses_by_route
+      |> Enum.split_with(fn %{route_id: route_id} -> route_id in @green_line_branches end)
+
+    consolidated_green_line_entries =
+      green_line_entries
+      |> Enum.group_by(& &1.statuses)
+      |> Enum.to_list()
+      |> convert_gl_branches_to_sub_routes()
+
+    other_entries ++ consolidated_green_line_entries
+  end
+
+  defp convert_gl_branches_to_sub_routes([{statuses, _}]) do
+    [
+      %{
+        route_id: "Green",
+        sub_routes: [],
+        statuses: statuses
+      }
+    ]
+  end
+
+  defp convert_gl_branches_to_sub_routes(entries) do
+    entries
+    |> Enum.map(fn {statuses, routes} ->
+      %{
+        route_id: "Green",
+        sub_routes: routes |> Enum.map(& &1.route_id),
+        statuses: statuses
+      }
+    end)
+  end
+
+  defp combine_mattapan_with_red_line(statuses_by_route) do
+    {mattapan_entries, other_entries} =
+      statuses_by_route
+      |> Enum.split_with(fn %{route_id: route_id} -> route_id == "Mattapan" end)
+
+    new_mattapan_entries =
+      case mattapan_entries do
+        [%{statuses: [%{description: "Normal Service"}]}] ->
+          []
+
+        _ ->
+          mattapan_entries
+          |> Enum.map(fn %{statuses: statuses} ->
+            %{route_id: "Red", sub_routes: ["Mattapan"], statuses: statuses}
+          end)
+      end
+
+    other_entries ++ new_mattapan_entries
+  end
+
+  defp sort_routes_and_sub_routes(entries) do
+    line_indexes = @lines |> Enum.with_index() |> Map.new()
+
+    entries
+    |> Enum.sort_by(fn %{
+                         route_id: route_id,
+                         sub_routes: sub_routes,
+                         statuses: [%{description: description} | _]
+                       } ->
+      description_sort_order =
+        case description do
+          "Normal Service" -> 1
+          _ -> 0
+        end
+
+      base_route_sort_order =
+        case sub_routes do
+          [] -> 0
+          _ -> 1
+        end
+
+      {
+        line_indexes |> Map.get(route_id),
+        base_route_sort_order,
+        description_sort_order,
+        sub_routes
+      }
+    end)
+  end
```

Comment:
This function is both long and deeply nested. It could probably be revisited.

Reply:
It's long, but is it deeply nested? I think it just has three fairly simple things going on that each take up a few lines of code.

Reply:
To me, it's deeply nested...as indicated by its horizontal shift. Within the function you have a sort_by and then two case statements.

Reply:
Extracted!